### PR TITLE
[#157340065] User can add an asset type

### PR DESCRIPTION
**Pull request by HawiCaesar**
## What does this PR do?
User can add asset type
## Description of Task to be completed?
- add component and container
- add async action to create asset type on the backend
- add reducer function to update frontend `store`
## How should this be manually tested?
`npm test` 
AND/OR
- Login
- Click on Assets on the Sidebar
- Click on the `+` icon next to the `Asset Types` column. A modal should pop up.
- Add the name of the asset type, select a sub category then click on save. The modal will close and a `success` message should appear on the top right corner of the page.
## What are the relevant pivotal tracker stories?
[157340065](https://www.pivotaltracker.com/story/show/157340065)
## Any background context you want to add?
n/a
## Important notes
n/a
## Packages installed
n/a
## Deployment note
n/a
## Related PRs branch | PR (branch_name) | (pr_link)
Branch | PR
------ | ------
other_pr_production | [link]()
other_pr_master | [link]()
## Todos
- [x] Raise PR
- [x] Test
- [ ] Documentation
## Screenshots
screen shot 2018-07-04 at 18 11 29

### CODE REVIEW

**Comment by sachungo** [comment]
@HawiCaesar coveralls still need more tests 🙂

File: src/__test__/actions/assetTypes.actions.test.js [hunk 1]
```diff
@@ -1,43 +1,82 @@
 // third-party libraries
 import expect from 'expect';
-import axios from 'axios';
-import MockAdapter from 'axios-mock-adapter';
 import thunk from 'redux-thunk';
 import configureMockStore from 'redux-mock-store';
+import moxios from 'moxios';
```
Comment (sachungo) [inline src/__test__/actions/assetTypes.actions.test.js]: @HawiCaesar why was this installed? Isn't `axios-mock-adapter` doing a similar job?

File: src/_actions/assetTypes.actions.js [hunk 2]
```diff
@@ -15,4 +22,31 @@ export const loadAssetTypes = () => (dispatch) => {
...
+
+export const createAssetTypeSuccess = assetType => (
+  { type: CREATE_ASSET_TYPE_SUCCESS, payload: assetType }
+);
+
+export const createAssetTypeFailure = error => (
+  { type: CREATE_ASSET_TYPE_FAILURE, payload: error }
+);
+
+export const updateToastMessageContent = (message, type) => (
```
Comment (sachungo) [inline src/_actions/assetTypes.actions.js]: Arthur centralized this in his PR, please liaise with him to know when he's merged to develop. You'll have to import `updateToastMessageContent`, not create it again

File: src/_actions/assetTypes.actions.js [hunk 3]
```diff
@@ -15,4 +22,31 @@ export const loadAssetTypes = () => (dispatch) => {
...
+    dispatch(createAssetTypeSuccess(response.data));
+    dispatch(updateToastMessageContent('Asset Type Saved Successfully',
+      'success'));
+  }).catch((error) => {
+    dispatch(createAssetTypeFailure(error));
+    dispatch(updateToastMessageContent('Could Not Save The Asset Type',
+      'error'));
+  });
+
+export const createAssetTypeSuccess = assetType => (
```
Comment (sachungo) [inline src/_actions/assetTypes.actions.js]: This is a misuse of `export` keyword. Only export action creators when they are being imported in a different file. The same comment applies to all action creators in this PR

File: src/_actions/subcategory.actions.js [hunk 4]
```diff
@@ -8,15 +8,10 @@ const { LOAD_SUBCATEGORIES_SUCCESS } = constants;
...
-  return ((dispatch) => {
-    return axios.get('https://my-json-server.typicode.com/HawiCaesar/jsonplaceholders-demo/subcategories').then((response) => {
-      dispatch(loadSubCategoriesSuccess(response.data));
-    }).catch((error) => {
-      console.log(error);
-    });
-  });
-}
+export const loadSubCategories = () => (
+  dispatch => axios.get('asset-sub-categories').then((response) => {
```
Comment (sachungo) [inline src/_actions/subcategory.actions.js]: What happens when the request errors?

File: src/_components/AssetTypes/AssetTypesContainer.jsx [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+
+import AssetTypesComponent from '../../components/AssetTypes/AssetTypesComponent';
+import { ToastMessage } from '../../_utils/ToastMessage';
+
+import { loadSubCategories } from '../../_actions/subcategory.actions';
+import { createAssetType } from '../../_actions/assetTypes.actions';
+import resetToastMessageContent from '../../_actions/resetToastMessage.actions';
+
+export class AssetTypesContainer extends React.Component {
+  constructor(props) {
```
Comment (sachungo) [inline src/_components/AssetTypes/AssetTypesContainer.jsx]: The constructor is unnecessary as no state is initialized with `props`. Declare the state as a class variable, i.e.
```
export class AssetTypesContainer extends React.Component {
  state = {
    assetType: '',
    ...
  };
  ...
}

File: src/components/AssetTypes/AssetTypesComponent.jsx [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+
+const placeMakesInSemanticUIOptions = props =>
+  props.map((option, index) => ({
+    key: index,
+    text: option.sub_category_name,
+    value: option.id
+  }));
+
+const AssetTypesComponent = props => (
+  <div>
```
Comment (sachungo) [inline src/components/AssetTypes/AssetTypesComponent.jsx]: `<div>` is unnecessary as `<form>` encapsulates all the markup

File: src/components/common/TextInputComponent.jsx [hunk 7]
```diff
@@ -2,26 +2,26 @@ import React from 'react';
...
-const TextInputComponent = (props) => (
+const TextInputComponent = props => (
   <Input
     fluid
     size="large"
-    style={{width: "65%"}}
+    style={{ width: '65%' }}
     icon="pencil"
-    placeholder={props.placeholder}
+    placeholder={props.placeHolder}
```
Comment (sachungo) [inline src/components/common/TextInputComponent.jsx]: I hope you have verified that this change doesn't affect existing functionality/UI

File: src/__test__/actions/assetTypes.actions.test.js (hunk 1 above)
Comment (HawiCaesar) [reply src/__test__/actions/assetTypes.actions.test.js]: @sachungo I installed it but it was actually going to the url. Moxios does a better of job of testing and testing failure scenarios.

File: src/_actions/assetTypes.actions.js (hunk 2 above)
Comment (HawiCaesar) [reply src/_actions/assetTypes.actions.js]: Ok, I will liaise with him.

File: src/_components/AssetTypes/AssetTypesContainer.jsx (hunk 5 above)
Comment (HawiCaesar) [reply src/_components/AssetTypes/AssetTypesContainer.jsx]: Alright.

File: src/components/AssetTypes/AssetTypesComponent.jsx (hunk 6 above)
Comment (HawiCaesar) [reply src/components/AssetTypes/AssetTypesComponent.jsx]: I will change this.

File: src/components/common/TextInputComponent.jsx (hunk 7 above)
Comment (HawiCaesar) [reply src/components/common/TextInputComponent.jsx]: Yes I have verified the change does not affect existing functionality/UI.

File: src/_actions/subcategory.actions.js (hunk 4 above)
Comment (HawiCaesar) [reply src/_actions/subcategory.actions.js]: 😄 I will add the catch snippet.

File: src/_actions/assetTypes.actions.js (hunk 3 above)
Comment (HawiCaesar) [reply src/_actions/assetTypes.actions.js]: Alright makes sense.

File: src/__test__/actions/assetTypes.actions.test.js (hunk 1 above)
Comment (sachungo) [reply src/__test__/actions/assetTypes.actions.test.js]: noted

File: src/_actions/assetTypes.actions.js (hunk 2 above)
Comment (sachungo) [reply src/_actions/assetTypes.actions.js]: @HawiCaesar this hasn't been deleted yet

File: src/_components/AssetTypes/AddAssetTypesContainer.jsx [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+import { ToastMessage } from '../../_utils/ToastMessage';
+
+import { loadSubCategories } from '../../_actions/subcategory.actions';
+import { createAssetType } from '../../_actions/assetTypes.actions';
+import resetToastMessageContent from '../../_actions/toastMessage.actions';
+
+export class AddAssetTypesContainer extends React.Component {
+  constructor(props) {
+    super(props);
+    this.state = {
```
Comment (sachungo) [inline src/_components/AssetTypes/AddAssetTypesContainer.jsx]: The constructor use is more ideal if binding methods or initializing state using a prop value. Therefore, declare the state as a class variable. I.e.

*Necessary use of constructor*
```
export class AddAssetTypesContainer extends React.Component {
  constructor(props) {
    this.state = {
      myState: props.myProp
    }
    this.handleFunction = this.handleFunction.bind(this);
  }
  ...
  ...
}
```
*Note*: Use of [props to initialize state is usually discouraged](https://reactjs.org/docs/react-component.html#constructor), unless you want to ignore the prop update

*State declared as a variable*
```
export class AddAssetTypesContainer extends React.Component {
  state = {
     myState: 0
  };
  ...
  ...
  render() {...}
}
```

Comment (HawiCaesar) [reply src/_components/AssetTypes/AddAssetTypesContainer.jsx]: @sachungo oops I had forgotten to remove the unnecessary `updateToastMessageContent` in `assetTypes.actions.js`.

I see the point when to use and not use the constructor. Adjusting the code.

**Pull request merged**